README.md
# EasyPost Python Client Library

EasyPost is a simple shipping API. You can sign up for an account at https://easypost.com

Requirements
------------

* [Requests](http://docs.python-requests.org/en/latest/) (if not on Google App Engine)

Installation
------------

You can install easypost via pip with:

    pip install easypost

Alternatively, you can clone the EasyPost python client repository:

    git clone https://github.com/EasyPost/easypost-python

Install:

    python setup.py install

Import the EasyPost client:

    import easypost

Example
-------

```python
import easypost
easypost.api_key = 'cueqNZUb3ldeWTNX7MU3Mel8UXtaAMUi'

# create and verify addresses
to_address = easypost.Address.create(
  verify=["delivery"],
  name = "Dr. Steve Brule",
  street1 = "179 N Harbor Dr",
  street2 = "",
  city = "Redondo Beach",
  state = "CA",
  zip = "90277",
  country = "US",
  phone = "310-808-5243"
)
from_address = easypost.Address.create(
  verify=["delivery"],
  name = "EasyPost",
  street1 = "118 2nd Street",
  street2 = "4th Floor",
  city = "San Francisco",
  state = "CA",
  zip = "94105",
  country = "US",
  phone = "415-456-7890"
)

# create parcel
try:
  parcel = easypost.Parcel.create(
    predefined_package = "Parcel",
    weight = 21.2
  )
except easypost.Error as e:
  print e.message
  if e.param != None:
    print 'Specifically an invalid param: ' + e.param

try:
  parcel = easypost.Parcel.create(
    length = 10.2,
    width = 7.8,
    height = 4.3,
    weight = 21.2
  )
except easypost.Error as e:
  raise e

# create customs_info form for intl shipping
customs_item = easypost.CustomsItem.create(
  description = "EasyPost t-shirts",
  hs_tariff_number = 123456,
  origin_country = "US",
  quantity = 2,
  value = 96.27,
  weight = 21.1
)
customs_info = easypost.CustomsInfo.create(
  customs_certify = 1,
  customs_signer = "Hector Hammerfall",
  contents_type = "gift",
  contents_explanation = "",
  eel_pfc = "NOEEI 30.37(a)",
  non_delivery_option = "return",
  restriction_type = "none",
  restriction_comments = "",
  customs_items = [customs_item]
)

# create shipment
shipment = easypost.Shipment.create(
  to_address = to_address,
  from_address = from_address,
  parcel = parcel,
  customs_info = customs_info
)

# buy postage label with one of the rate objects
shipment.buy(rate = shipment.rates[0])
# alternatively: shipment.buy(rate = shipment.lowest_rate())

print shipment.tracking_code
print shipment.postage_label.label_url

# Insure the shipment for the value
shipment.insure(amount=100)

print shipment.insurance
```

Documentation
-------------

Up-to-date documentation at: https://www.easypost.com/docs

Tests
-----
Coming soon!


CHANGELOG
=== 3.5.2 2017-02-14

* Added get_rates method to Order objects


=== 3.5.1 2017-01-19

* Fixed create for ScanForms


=== 3.5.0 2017-01-18

* Added basic CRUD methods for Webhook Objects
* Fixed Order test


=== 3.4.0 2016-12-20

* Added session pooling


=== 3.3.0 2016-12-15

* Added support for Report objects


=== 3.2.2 2016-07-29

* Added support for Insurance objects


=== 3.2.1 2016-07-18

* Added to_json method for EasyPost object. Thanks ThePsyjo!


=== 3.2.0 2016-05-17

* Remove api_keys from object dumps
* Fixed address tests to keep them in line with the new API messages
* This was previously known as version 3.1.3


=== 3.1.2 2016-03-04

* Added a suite of unittest tests. Thanks wyounas!


=== 3.1.1 2016-02-12

* Added ability to interact with Users (create, retrieve and update)


=== 3.1.0 2015-12-08

* Add verifications to Address.create


=== 3.0.6 2015-11-24

* Added Blob.retrieve for fetching urls for blobs stored by EasyPost.


=== 3.0.5 2015-11-18

* Changed the interface on Tracker.create_list to return True rather than the
Trackers themselves


=== 3.0.4 2015-11-16

* Minor bugfix


=== 3.0.3 2015-11-16

* Added a Tracker.all_updated method for retrieving a large number of Trackers
by status or tracking_details updated datetime


=== 3.0.2 2015-11-10

* Added a Tracker.create_list method for creating a large number of Trackers
at once


=== 3.0.1 2015-11-04

* Fixed some examples and added some new ones, particularly dealing with Tracker.all


=== 3.0.0 2015-10-19

* Fixed bug where retrieving a shipment by tracking code or reference doesn't set the ID properly


=== 2.0.16 2015-08-10

* Added ability to interact with carrier accounts (full CRUD)


=== 2.0.15 2015-07-31

* Fixed bug with address verification url rendering

=== 2.0.14 2015-03-30

* Fix numerous bugs, including Python3 encoding


=== 2.0.13 2015-01-09

* Added python3 support


=== 2.0.12 2014-11-04

* Added tracker to shipment response
* Added tracker example


=== 2.0.11 2014-09-19

* Added Order support.


=== 2.0.10 2014-09-10

* Added Pickup cancellation method.


=== 2.0.9 2014-09-02

* Added Pickup resource for scheduling pickups.


=== 2.0.6 2013-10-17

* Added Event resource for webhook digestion.
* Added buy method to Batch.


=== 2.0.5 2013-09-22

* Bug Fix: UTF-8 input handled more gracefully.


=== 2.0.4 2013-07-31

* API Addition: Tracker resource added. Trackers can be used to register any tracking code with EasyPost webhooks.


=== 2.0.3 2013-07-23

* API Addition: Shipment.track_with_code returns tracking details for any tracking code.


=== 2.0.2 2013-07-07

* Bug fix: address.create_and_verify is now a classmethod.


=== 2.0.1 2013-07-05

* Added function to Address to all creating and verifying at the same time.
* Add label function to Shipment to request specific label file_formats (pdf, epl2, zpl).
* Add insure function to Shipment. Add insurance to any shipment in one call!
* Fixed shipment.get_rates bug.


example.py
import easypost
easypost.api_key = 'cueqNZUb3ldeWTNX7MU3Mel8UXtaAMUi'
# easypost.api_base = 'http://localhost:5000/v2'

# create addresses
to_address = easypost.Address.create(
    name="Sawyer Bateman",
    street1="1A Larkspur Cres.",
    street2="",
    city="St. Albert",
    state="AB",
    zip="t8n2m4",
    country="CA",
    phone="780-283-9384"
)
from_address = easypost.Address.create(
    company="EasyPost",
    street1="118 2nd St",
    street2="4th Fl",
    city="San Francisco",
    state="CA",
    zip="94105",
    phone="415-456-7890"
)

# verify address
try:
    verified_from_address = from_address.verify()
except easypost.Error as e:
    raise e
if hasattr(verified_from_address, 'message'):
    # the from address is not invalid, but it has an issue
    pass

# create parcel
try:
    parcel = easypost.Parcel.create(
        predefined_package="InvalidPackageName",
        weight=21.2
    )
except easypost.Error as e:
    print(e)
    if e.param is not None:
        print('Specifically an invalid param: ' + e.param)

try:
    parcel = easypost.Parcel.create(
        length=10.2,
        width=7.8,
        height=4.3,
        weight=21.2
    )
except easypost.Error as e:
    raise e

# create customs_info form for intl shipping
customs_item = easypost.CustomsItem.create(
    description="EasyPost t-shirts",
    hs_tariff_number=123456,
    origin_country="US",
    quantity=2,
    value=96.27,
    weight=21.1
)
customs_info = easypost.CustomsInfo.create(
    customs_certify=1,
    customs_signer="Hector Hammerfall",
    contents_type="gift",
    contents_explanation="",
    eel_pfc="NOEEI 30.37(a)",
    non_delivery_option="return",
    restriction_type="none",
    restriction_comments="",
    customs_items=[customs_item]
)

# create shipment
shipment = easypost.Shipment.create(
    to_address=to_address,
    from_address=verified_from_address,
    parcel=parcel,
    customs_info=customs_info
)

# buy postage label with one of the rate objects
# shipment.buy(rate = shipment.rates[0])
# shipment.buy(rate = shipment.lowest_rate('usps'))
shipment.buy(rate=shipment.lowest_rate(
    ['USPS', 'ups'], 'priorityMAILInternational'))
# Insure the shipment for the value
shipment.insure(amount=100)

# shipment.refund()
# print(shipment.refund_status)

print(shipment.tracking_code)
print(shipment.insurance)
print(shipment.postage_label.label_url)
